### PR TITLE
Multi-threaded server and sync/async request forwarding

### LINKED ISSUE #1: Support concurrent requests
Option for supporting multi-threaded server-side to support concurrent requests

### LINKED ISSUE #8: Possibility to customize type of service-to-service interactions
We could have two different calls: 1. Synchronous calls or 2. Asynchronous calls.

Tasks are executed parallelly even in different microservices

### PR DESCRIPTION
**Pull request by alekodu**
Closes #1
Closes #8

Added:
* Multi-threaded support in Flask server
* Sync/async request forwarding for i/o-bounded tasks
* Update service description file and config map with additional configuration parameters to produce:
image

### CODE REVIEW

File: generator/src/pkg/generate/generate.go [hunk 1]
```diff
@@ -88,6 +90,7 @@ type Services struct {
 	Name      string      `json:"name"`
 	Clusters  []Clusters  `json:"clusters"`
 	Resources Resources   `json:"resources"`
+	Processes	int			`json:"processes"`
```
[the remark this answers is not in the document]
Comment (alekodu) [reply generator/src/pkg/generate/generate.go:93]: Here "processes" is added inside the "Service" struct, isn't it?

File: generator/src/pkg/generate/generate.go [hunk 2]
```diff
@@ -183,8 +191,13 @@ func Create(config Config, readinessProbe int, clusters []string) {
 			resources.Requests.Memory = requestsMemoryDefault
 		}
 
-		serv_endpoints := []Endpoints(config.Services[i].Endpoints)
-		serv_ep_json, err := json.Marshal(map[string][]Endpoints{"endpoints": serv_endpoints})
+		// TODO: add processes param
```
[the remark this answers is not in the document]
Comment (alekodu) [reply generator/src/pkg/generate/generate.go:194]: right, I forgot to remove this line

File: model/src/controller/routes.py [hunk 3]
```diff
@@ -16,14 +16,18 @@
...
+from aiohttp import ClientSession
+import asyncio
 
 simple_page = Blueprint("simple_page", __name__,)
 logger = logging.getLogger(__name__)
 
 # TODO: So far, we only support one endpoint per service...
-service_config = path.SERVICE_CONFIG[0]
+service_endpoint = path.SERVICE_CONFIG["endpoints"][0]
+#service_namespace = path.SERVICE_CONFIG["namespace"]
```
[the remark this answers is not in the document]
Comment (alekodu) [reply model/src/controller/routes.py:29]: This is the namespace on which this microservice is deployed... so far I am not using it but we may use it in the future... that is why I commented it

File: generator/input/new_description.json [hunk 4]
```diff
@@ -18,28 +18,32 @@
...
           "name": "/end-1",
           "protocol": "http",
           "cpu_consumption": 0.003,
           "network_consumption": 0.002,
           "memory_consumption": 0.003,
+          "forward_requests": "asynchronous",
           "calledservices": [
             {
               "cluster": "cluster-1",
+              "namespace": "ns-2",
```
[the remark this answers is not in the document]
Comment (alekodu) [reply generator/input/new_description.json:41]: you are right, it is not strictly needed but when wanting to route requests to a particular cluster and/or namespace. However, I realized that such cluster/namespace-based routing doesn't need to be configured at the application... It should be configured with Istio traffic policies. So, I will remove both cluster and namespace from here.

File: model/src/controller/routes.py [hunk 5]
```diff
@@ -41,19 +45,35 @@ def getForwardHeaders(request):
...
+    if service_endpoint["forwardRequests"] == "asynchronous":
+        async with ClientSession() as session:
+            # TODO: CPU-bounded tasks not supported yet
+            io_tasks = []
+            for svc in service_endpoint["calledServices"]:
+                io_task = asyncio.create_task(execute_io_bounded_task(session=session, target_service=svc, forward_headers=headers))
+                io_tasks.append(io_task)
+            services = await asyncio.gather(*io_tasks)
+
+        # Generate the HTML response
```
[the remark this answers is not in the document]
Comment (alekodu) [reply model/src/controller/routes.py:60]: Yes, I will jsonify this response data

**Review by salehsedghpour — APPROVED** [review]

**Pull request merged**